
--- a/src/lib-storage/index/index-storage.c	Wed Jan 14 13:55:04 2009 -0500
+++ b/src/lib-storage/index/index-storage.c	Tue Jan 27 14:31:18 2009 -0500
@@ -467,7 +467,7 @@ bool index_storage_allow_new_keywords(st
 bool index_storage_allow_new_keywords(struct mailbox *box)
 {
 	/* FIXME: return FALSE if we're full */
-	return index_storage_is_readonly(box);
+	return !index_storage_is_readonly(box);
 }
 
 bool index_storage_is_inconsistent(struct mailbox *box)




