--- dovecot-2.2.35/src/lib-master/master-service-settings-cache.c	2018-03-21 10:15:09.097480691 +0100
+++ dovecot-2.2.35/src/lib-master/master-service-settings-cache.c~	2018-03-19 10:30:01.000000000 +0100
@@ -131,7 +131,7 @@ match_local_name(const char *local_name,
 			return TRUE;
 		local_name = ptr+1;
 	}
-	return dns_match_wildcard(filter_local_name, local_name) == 0;
+	return dns_match_wildcard(local_name, filter_local_name) == 0;
 }
 
 /* Remove any elements which there is no filter for */
