--- dovecot-2.2.18/Makefile.am~	2015-02-19 12:44:40.000000000 +0100
+++ dovecot-2.2.18/Makefile.am	2015-05-17 08:43:25.681245560 +0200
@@ -64,7 +64,7 @@ install-exec-hook:
 	grep -v '^LIBDOVECOT_.*_INCLUDE' dovecot-config | \
 	grep -v '^LIBDOVECOT.*_DEPS' | sed \
 	-e "s|^\(DOVECOT_INSTALLED\)=.*$$|\1=yes|" \
-	-e "s|^\(LIBDOVECOT\)=.*$$|\1='-L$(pkglibdir) -ldovecot'|" \
+	-e "s|^\(LIBDOVECOT\)=.*$$|\1='-rpath $(pkglibdir) -L$(pkglibdir) -ldovecot'|" \
 	-e "s|^\(LIBDOVECOT_LOGIN\)=.*$$|\1='-ldovecot-login $(SSL_LIBS)'|" \
 	-e "s|^\(LIBDOVECOT_SQL\)=.*$$|\1=-ldovecot-sql|" \
 	-e "s|^\(LIBDOVECOT_COMPRESS\)=.*$$|\1=-ldovecot-compression|" \
