--- dovecot-2.2.27/doc/example-config/conf.d/10-ssl.conf~	2016-12-03 21:37:57.000000000 +0100
+++ dovecot-2.2.27/doc/example-config/conf.d/10-ssl.conf	2016-12-03 21:38:42.082007401 +0100
@@ -47,9 +47,10 @@ ssl = no
 
 # SSL protocols to use
 #ssl_protocols = !SSLv3
+ssl_protocols = !SSLv2 !SSLv3
 
 # SSL ciphers to use
-#ssl_cipher_list = ALL:!LOW:!SSLv2:!EXP:!aNULL
+ssl_cipher_list = ALL:!LOW:!SSLv2:!EXP:!aNULL:!SSLv3
 
 # Prefer the server's order of ciphers over client's.
 #ssl_prefer_server_ciphers = no
