From 6c969ac21a43cc10ee1f1a91a4f39e4864c886cb Mon Sep 17 00:00:00 2001
From: Aki Tuomi <aki.tuomi@dovecot.fi>
Date: Fri, 15 Jul 2016 11:31:25 +0300
Subject: [PATCH] auth: Remove i_assert for credentials scheme

---
 src/auth/auth-request.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/src/auth/auth-request.c b/src/auth/auth-request.c
index 15cffeb..c701940 100644
--- a/src/auth/auth-request.c
+++ b/src/auth/auth-request.c
@@ -1065,8 +1065,6 @@ void auth_request_lookup_credentials(struct auth_request *request,
 
 	if (request->credentials_scheme == NULL)
 		request->credentials_scheme = p_strdup(request->pool, scheme);
-	else
-		i_assert(request->credentials_scheme == scheme);
 
 	if (request->policy_processed)
 		auth_request_lookup_credentials_policy_continue(request, callback);
